crates/wasm-shrink/src/lib.rs

```diff
@@ -0,0 +1,392 @@
+    #[cfg_attr(feature = "structopt", structopt(short, long, default_value = "info"))]
+    #[doc(hidden)]
+    pub log_level: LogLevel,
+
+    /// The interestingness predicate script.
+    predicate: PathBuf,
+
+    /// The input Wasm.
+    input: PathBuf,
+}
+
+#[doc(hidden)]
+pub enum LogLevel {
+    Off,
+    Error,
+    Warn,
+    Info,
+    Debug,
+    Trace,
+}
+
+impl std::str::FromStr for LogLevel {
+    type Err = String;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        match s {
+            "off" => Ok(LogLevel::Off),
+            "error" => Ok(LogLevel::Error),
+            "warn" => Ok(LogLevel::Warn),
+            "info" => Ok(LogLevel::Info),
+            "debug" => Ok(LogLevel::Debug),
+            "trace" => Ok(LogLevel::Trace),
+            _ => Err(format!("Not a valid log level: {}", s)),
+        }
+    }
+}
+
+impl WasmShrink {
+    /// Construct a new instance of `WasmShrink` with the given predicate script
+    /// and input Wasm file.
+    pub fn new(predicate: impl Into<PathBuf>, input_wasm: impl Into<PathBuf>) -> WasmShrink {
+        WasmShrink {
+            output: None,
+            attempts: 1_000,
+            seed: 42,
+            log_level: LogLevel::Info,
+            predicate: predicate.into(),
+            input: input_wasm.into(),
+        }
+    }
+
+    /// The output file path to write the shrunken Wasm file to.
+    ///
+    /// By default, a file path based on the input will be generated.
+    pub fn output(mut self, output: impl Into<PathBuf>) -> WasmShrink {
+        self.output = Some(output.into());
+        self
+    }
+
+    /// The number of shrink attempts to try before considering a Wasm module as
+    /// small as it will ever get.
+    pub fn attempts(mut self, attempts: u32) -> WasmShrink {
+        self.attempts = attempts;
+        self
+    }
+
+    /// Run this configured Wasm shrinking task.
+    ///
+    /// Returns information and metrics about the shrink task, such as the
+    /// shrunken Wasm's output file path, the size of the input Wasm, the size
+    /// of the output Wasm, etc.
+    pub fn run(self) -> Result<ShrinkInfo> {
+        // Prerequisites for the predicate.
+        anyhow::ensure!(
+            self.predicate.is_file(),
+            "The predicate script '{}' does not exist.",
+            self.predicate.display()
+        );
+        anyhow::ensure!(
+            self.predicate.is_executable(),
+            "The predicate script '{}' is not executable.",
+            self.predicate.display()
+        );
+
+        // Prerequisites for the input Wasm.
+        anyhow::ensure!(
+            self.input.is_file(),
+            "The input Wasm file '{}' does not exist.",
+            self.input.display()
+        );
+        anyhow::ensure!(
+            self.is_valid_wasm(&self.input)?,
+            "The input Wasm file '{}' is not valid Wasm.",
+            self.input.display(),
+        );
```

Comment:
Is this required? I would imagine that reduction could sometimes be useful for even invalid wasm files, e.g. in a wasm file that panics in the validator or something like that.

Reply:
Unfortunately, `wasm-mutate` won't really work on invalid Wasm files. We could fix that, but until then I think this should stay.